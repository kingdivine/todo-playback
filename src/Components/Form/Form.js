import React, { useState } from "react";

import "./Form.scss";

const Form = ({ todo, handleSave }) => {
  const [name, setName] = useState(todo ? todo.name : "");
  const [description, setDescription] = useState(todo ? todo.description : "");

  const handleNameInputKeyPress = (event) => {
    event.preventDefault();
    setName(event.target.value);
  };

  const handleDescriptionInputKeyPress = (event) => {
    event.preventDefault();
    setDescription(event.target.value);
  };

  const handleSaveClick = () => {
    handleSave(todo ? todo : null, name.trim(), description.trim());
  };

  return (
    <div className={"form-container"}>
      <div>
        <input
          value={name}
          type="text"
          placeholder="Name"
          maxLength="30"
          onChange={(e) => handleNameInputKeyPress(e)}
        ></input>
      </div>
      <div>
        <textarea
          value={description}
          placeholder="Enter task description..."
          maxLength="140"
          rows="10"
          onChange={(e) => handleDescriptionInputKeyPress(e)}
        ></textarea>
      </div>
      <div>
        <button
          className="save-btn"
          disabled={name.trim() === ""}
          onClick={handleSaveClick}
        >
          Save
        </button>
      </div>
    </div>
  );
};

export default Form;
