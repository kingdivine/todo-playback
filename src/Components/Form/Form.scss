@import "../../styles/media";

.form-container {
  margin-top: 20px;
}

input {
  margin: 5px;
  padding: 5px;
  border: 1px solid gray;
  @include phone-screen {
    width: 45%;
  }
  @include tablet-screen {
    width: 30%;
  }
  @include desktop-screen {
    width: 20%;
  }
  @include ultrawide-screen {
    width: 15%;
  }
}

textarea {
  resize: none;
  padding: 5px;
  border: 1px solid gray;
  @include phone-screen {
    width: 50%;
  }
  @include tablet-screen {
    width: 35%;
  }
  @include desktop-screen {
    width: 25%;
  }
  @include ultrawide-screen {
    width: 20%;
  }
}

.save-btn {
  background-color: #00a9c7;
}
