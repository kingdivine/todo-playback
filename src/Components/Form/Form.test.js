import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";

import Form from "./Form";

afterEach(cleanup);

describe("<Form/>", () => {
  const todo = {
    id: "abc123",
    name: "Item 1",
    description: "This is the first item to complete",
    createdDate: new Date("Thu Apr 09 2020 16:36:07 GMT+0100 (BST)"),
  };
  test("renders empty form correctly", () => {
    expect(
      render(<Form todo={null} handleSave={jest.fn()} />)
    ).toMatchSnapshot();
  });
  test("renders edit form correctly ", () => {
    expect(
      render(<Form todo={todo} handleSave={jest.fn()} />)
    ).toMatchSnapshot();
  });
  test("renders disbaled save button initially when empty", () => {
    const saveFn = jest.fn();
    const { getByText } = render(<Form todo={null} handleSave={saveFn} />);
    fireEvent.click(getByText("Save"));
    expect(saveFn).toHaveBeenCalledTimes(0);
  });
  test("enables save button after valid input", () => {
    const saveFn = jest.fn();
    const { getByText, getByPlaceholderText } = render(
      <Form todo={null} handleSave={saveFn} />
    );
    fireEvent.change(getByPlaceholderText("Name"), {
      target: { value: todo.name },
    });
    fireEvent.click(getByText("Save"));
    expect(saveFn).toHaveBeenCalledTimes(1);
  });
});
