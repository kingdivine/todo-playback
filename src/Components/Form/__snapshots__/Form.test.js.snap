// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> renders edit form correctly  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-container"
      >
        <div>
          <input
            maxlength="30"
            placeholder="Name"
            type="text"
            value="Item 1"
          />
        </div>
        <div>
          <textarea
            maxlength="140"
            placeholder="Enter task description..."
            rows="10"
          >
            This is the first item to complete
          </textarea>
        </div>
        <div>
          <button
            class="save-btn"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-container"
    >
      <div>
        <input
          maxlength="30"
          placeholder="Name"
          type="text"
          value="Item 1"
        />
      </div>
      <div>
        <textarea
          maxlength="140"
          placeholder="Enter task description..."
          rows="10"
        >
          This is the first item to complete
        </textarea>
      </div>
      <div>
        <button
          class="save-btn"
        >
          Save
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Form/> renders empty form correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-container"
      >
        <div>
          <input
            maxlength="30"
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div>
          <textarea
            maxlength="140"
            placeholder="Enter task description..."
            rows="10"
          />
        </div>
        <div>
          <button
            class="save-btn"
            disabled=""
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-container"
    >
      <div>
        <input
          maxlength="30"
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <div>
        <textarea
          maxlength="140"
          placeholder="Enter task description..."
          rows="10"
        />
      </div>
      <div>
        <button
          class="save-btn"
          disabled=""
        >
          Save
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
